File: backend/src/modules/credit/credit.service.js
// ============================================================
// ARTH Backend — Credit Service
// Prepares behavioral data and calls ML for credit scoring.
// ============================================================

const mlClient = require('../../integrations/mlClient');
const transactionService = require('../transaction/transaction.service');
const logger = require('../../utils/logger');

/**
 * Get credit score for a user.
 */
const getScore = async (userId, inputData = {}) => {
  const summary = await transactionService.getTransactionSummary(userId);

  const payload = {
    income: inputData.income || summary.totalIncome,
    expenses: inputData.expenses || summary.totalExpenses,
    transaction_count: summary.transactionCount,
    net_balance: summary.netBalance,
    // Behavioral signals
    expense_to_income_ratio: summary.totalIncome > 0
      ? summary.totalExpenses / summary.totalIncome
      : 1,
    category_diversity: Object.keys(summary.categoryBreakdown).length,
    avg_transaction_amount: summary.transactionCount > 0
      ? (summary.totalIncome + summary.totalExpenses) / summary.transactionCount
      : 0,
  };

  logger.info(`Credit score request for user ${userId}`);

  const result = await mlClient.getCreditScore(payload);

  return {
    score: result,
    financialSummary: {
      totalIncome: summary.totalIncome,
      totalExpenses: summary.totalExpenses,
      netBalance: summary.netBalance,
    },
  };
};

/**
 * Get credit score explanation for a user.
 */
const getExplanation = async (userId, inputData = {}) => {
  const summary = await transactionService.getTransactionSummary(userId);

  const payload = {
    income: inputData.income || summary.totalIncome,
    expenses: inputData.expenses || summary.totalExpenses,
    transaction_count: summary.transactionCount,
    net_balance: summary.netBalance,
    expense_to_income_ratio: summary.totalIncome > 0
      ? summary.totalExpenses / summary.totalIncome
      : 1,
    category_diversity: Object.keys(summary.categoryBreakdown).length,
    avg_transaction_amount: summary.transactionCount > 0
      ? (summary.totalIncome + summary.totalExpenses) / summary.transactionCount
      : 0,
  };

  logger.info(`Credit explanation request for user ${userId}`);

  const result = await mlClient.getCreditExplanation(payload);

  return {
    explanation: result,
    financialSummary: {
      totalIncome: summary.totalIncome,
      totalExpenses: summary.totalExpenses,
      categoryBreakdown: summary.categoryBreakdown,
    },
  };
};

module.exports = { getScore, getExplanation };

File: ml-service/app/services/credit_service.py
# ============================================================
# ARTH ML Service — Credit Service
# ============================================================

from app.models.credit_model import credit_model
from app.schemas.credit import CreditRequest
import logging

logger = logging.getLogger("arth.ml.credit.service")


def get_score(request: CreditRequest) -> dict:
    """Calculate credit score."""
    try:
        logger.info(f"Calculating credit score: income={request.income}")

        features = {
            "income": request.income,
            "expenses": request.expenses,
            "transaction_count": request.transaction_count,
            "expense_to_income_ratio": request.expense_to_income_ratio,
            "category_diversity": request.category_diversity,
            "net_balance": request.net_balance,
            "avg_transaction_amount": request.avg_transaction_amount,
        }

        result = credit_model.score(features)
        logger.info(f"Credit score result: {result['score']} (Grade {result['grade']})")
        return result

    except Exception as e:
        logger.error(f"Credit scoring failed: {str(e)}")
        raise


def get_explanation(request: CreditRequest) -> dict:
    """Get detailed credit explanation."""
    try:
        logger.info(f"Generating credit explanation for income={request.income}")

        features = {
            "income": request.income,
            "expenses": request.expenses,
            "transaction_count": request.transaction_count,
            "expense_to_income_ratio": request.expense_to_income_ratio,
            "category_diversity": request.category_diversity,
            "net_balance": request.net_balance,
            "avg_transaction_amount": request.avg_transaction_amount,
        }

        result = credit_model.explain(features)
        logger.info(f"Credit explanation generated: score={result['score']}")
        return result

    except Exception as e:
        logger.error(f"Credit explanation failed: {str(e)}")
        raise
